prov/rxm/src/rxm.h

```diff
@@ -239,14 +238,49 @@ struct rxm_recv_entry {
 };
 DECLARE_FREESTACK(struct rxm_recv_entry, rxm_recv_fs);
 
+#define RXM_QUEUE_HDR 		\
+	size_t 		size;	\
+	ofi_atomic32_t 	used;
+
+struct rxm_queue {
+	RXM_QUEUE_HDR
+};
+
+struct rxm_send_queue {
+	RXM_QUEUE_HDR
+	struct rxm_txe_fs *fs;
+	struct ofi_key_idx tx_key_idx;
+};
+
 struct rxm_recv_queue {
-	struct rxm_recv_fs *recv_fs;
+	RXM_QUEUE_HDR
+	struct rxm_recv_fs *fs;
 	struct dlist_entry recv_list;
 	struct dlist_entry unexp_msg_list;
 	dlist_func_t *match_recv;
 	dlist_func_t *match_unexp;
 };
 
+#define RXM_DEF_QUEUE_FUNC(q_type, e_type, name)		\
+static inline e_type *rxm_get_ ## name ## _entry(q_type *queue)	\
+{								\
+	if (ofi_atomic_get32(&queue->used) >= queue->size)	\
+		return NULL;					\
+								\
+	ofi_atomic_inc32(&queue->used);				\
+	return freestack_pop(queue->fs);			\
```

Comment:
freestack rounds up size to next power of 2 so isempty() isn't helpful. I'll look into cirq to check if they can be used instead.